### PR TITLE
refactor [JenkinsTests JenkinsCoverage], also run [JenkinsBuild]

### PR DESCRIPTION
**Pull request by calebcartwright**
Refs #2863 

Summary of changes:
* JenkinsTests badge refactored to new service arch. 
  * The badge message was also changed to match our other Test badges (details inline below)
  * Added a couple helpers for test-specific badge services with the intent that they be reused for our other test badges (currently Appveyor and Azure DevOps, CircleCI and Sonar in the future, etc.)
* JenkinsCoverage badge was refactored to leverage the new Jenkins related content added here and in #3302. 
  * I believe this also helps align more closely to the objectives in #3174  
  * Service tests updated to use Jenkins instances that appear to be more stable/performant
* The 3 Jenkins routes (Build, Coverage, and Tests) have been updated from their original shorthand/aliases to more explicit route names (with redirects to maintain backwards compatibility):
  * `/jenkins/s` --> `/jenkins/build`
  * `/jenkins/c` --> `/jenkins/coverage`
  * `/jenkins/t` --> `/jenkins/tests`

I know there's still ongoing discussion around what our url conventions are for `build`, but using `build`, `coverage`, and `tests` for Jenkins aligns the url's with the other service (Azure DevOps) we provide multiple metric badges for. 

Using `build` also seems to be _relatively_ common in our badge url's for other services where the service name doesn't automatically default to build status (like we do for TravisCI). Definitely open to feedback/suggestions though!

### CODE REVIEW

File: services/test-results.js [hunk 1]
```diff
@@ -84,8 +84,31 @@ function renderTestResultBadge({
   return { message, color }
 }
 
+const getDocumentation = ({ route }) => `
```
Comment (calebcartwright) [inline services/test-results.js]: This is one of those things that I saw duplicated in [appveyor-tests.service.js](https://github.com/badges/shields/blob/master/services/appveyor/appveyor-tests.service.js#L9) and [azure-devops-tests.service.js](https://github.com/badges/shields/blob/master/services/azure-devops/azure-devops-tests.service.js#L32). 

Instead of replicating it again for JenkinsTests, I figured it would be better to put it here in one place and consume it from the other services (if there's no opposition to this, I will open a future PR to update Appveyor and Azure DevOps test services to leverage this). We've also got some additional test badges that have been requested (for example CircleCI and Sonar) that can re-use this as well.

File: services/test-validators.js [hunk 2]
```diff
@@ -91,6 +91,39 @@ const isDependencyState = withRegex(
   /^(\d+ out of date|\d+ deprecated|up to date)$/
 )
 
+const getTestTotalsValidator = ({ passed, failed, skipped }) =>
```
Comment (calebcartwright) [inline services/test-validators.js]: Same comment as above, as both Appveyor and Azure DevOps Tests were replicating this same functional code in multiple places.

Instead of replicating it again for JenkinsTests, I figured it would be better to put it here in one place and consume it from the other service testers (if there's no opposition to this, I will open a future PR to update Appveyor and Azure DevOps test services to leverage this). We've also got some additional test badges that have been requested (for example CircleCI and Sonar) that can re-use this as well.

File: services/jenkins/jenkins-tests.service.js [hunk 3]
```diff
@@ -1,110 +1,154 @@
...
+    passed,
+    failed,
+    skipped,
+    total,
+    passedLabel,
+    failedLabel,
+    skippedLabel,
+    isCompact,
+  }) {
+    return renderTestResultBadge({
```
Comment (calebcartwright) [inline services/jenkins/jenkins-tests.service.js]: This is a clear change to our badge message for Jenkins Tests (before the refactor the message was: `${passed} / ${total}`.

However, I believe this is worthwhile as reusing the `renderTestResultBadge` helper here provides a lot of new formatting capabilities and brings the Jenkins Test badge message format in alignment with our other test badges (Appveyor and Azure DevOps)

File: services/jenkins/jenkins-coverage.tester.js [hunk 4]
```diff
@@ -1,176 +1,27 @@
...
-const t = new ServiceTester({
-  id: 'jenkins-coverage',
-  title: 'JenkinsCoverage',
-  pathPrefix: '/jenkins',
-})
-module.exports = t
+// The below page includes links to various publicly accessible Jenkins instances
+// although many of the links are dead, it is is still a helpful resource for finding
+// target Jenkins instances/jobs to use for testing.
+// https://wiki.jenkins.io/pages/viewpage.action?pageId=58001258
```
Comment (calebcartwright) [inline services/jenkins/jenkins-coverage.tester.js]: I've updated the tests to use different Jenkins instances that seem to be much more performant and stable than the Apache instance. That gives us good test targets to validate both jacoco and cobertura coverage formats, so I've removed several of the mock tests accordingly (I believe we were mocking due to the lack of consistently available live jobs)

File: services/jenkins/jenkins-coverage.service.js [hunk 5]
```diff
@@ -1,55 +1,60 @@
...
+              name: Joi.string().required(),
+              ratio: Joi.number()
+                .min(0)
+                .max(100)
+                .required(),
+            })
+          )
+          .has(Joi.object({ name: 'Lines' }))
+          .min(1)
+          .required(),
```
Comment (calebcartwright) [inline services/jenkins/jenkins-coverage.service.js]: Jenkins will return an HTTP 404 if a job is requested that did not publish code coverage in the cobertura format (which is why on a 404 we use the error message `job or coverage not found`). 

If the target job has cobertura coverage then the elements array will include an object with a `Lines` field, which is what we need.

File: services/jenkins/jenkins-build-redirect.service.js [hunk 6]
```diff
@@ -6,11 +6,21 @@ module.exports = [
...
+    dateAdded: new Date('2019-04-20'),
+  }),
+  redirector({
+    category: 'build',
+    route: {
+      base: 'jenkins/s',
+      pattern: ':protocol(http|https)/:host+/:job+',
+    },
+    transformPath: ({ protocol, host, job }) =>
+      `/jenkins/build/${protocol}/${host}/${job}`,
```
Comment (paulmelnikow) [inline services/jenkins/jenkins-build-redirect.service.js]: It's not immediately obvious these are the same. Maybe you could define the `transformPath` function above to help clarify?

File: services/jenkins/jenkins-coverage-redirector.service.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+'use strict'
+
+const { redirector } = require('..')
+
+module.exports = [
```
Comment (paulmelnikow) [inline services/jenkins/jenkins-coverage-redirector.service.js]: Since there's only one, you can `module.exports = redirector({ ... })` and then you can use `createServiceTester()`.

File: services/jenkins/jenkins-coverage.tester.js [hunk 8]
```diff
@@ -1,176 +1,27 @@
...
```
Comment (paulmelnikow) [inline services/jenkins/jenkins-coverage.tester.js]: 🙌

File: services/jenkins/jenkins-tests.service.js [hunk 9]
```diff
@@ -1,110 +1,154 @@
...
+  }
 
-        if (serverSecrets.jenkins_user) {
-          options.auth = {
-            user: serverSecrets.jenkins_user,
-            pass: serverSecrets.jenkins_pass,
-          }
-        }
+  transform({ json }) {
+    const testsObject = json.actions.find(o => o.hasOwnProperty('failCount'))
```
Comment (paulmelnikow) [inline services/jenkins/jenkins-tests.service.js]: `.hasOwnProperty` is a pretty archaic form that I don't think we need to use anymore.

```suggestion
    const testsObject = json.actions.find(o => 'failCount' in o)
```

File: services/test-results.js (hunk 1 above)
Comment (paulmelnikow) [reply services/test-results.js]: Sounds good! I'm not sure it's essential to put the route in there, and leaving it out would simplify this, though I don't object.

File: services/test-validators.js (hunk 2 above)
Comment (paulmelnikow) [reply services/test-validators.js]: Looks nice! Appreciate the cleanup, definitely makes this more enjoyable to read.

I've a slight preference to call these `makeTestTotalsValidator`.

File: services/validators.js [hunk 10]
```diff
@@ -2,11 +2,14 @@
...
+  .min(0)
+
 module.exports = {
-  nonNegativeInteger: Joi.number()
-    .integer()
-    .min(0)
-    .required(),
+  optionalNonNegativeInteger,
+
+  nonNegativeInteger: optionalNonNegativeInteger.required(),
```
Comment (paulmelnikow) [inline services/validators.js]: Bet there are a lot of places where we can use this!

File: services/jenkins/jenkins-build-redirect.service.js (hunk 6 above)
Comment (calebcartwright) [reply services/jenkins/jenkins-build-redirect.service.js]: I'm not following, could you rephrase? 

The legacy route accepted a base of `jenkins|jenkins-ci` and for the metric it accepted `s` as a shorthand for build status. Are you suggesting combining that into a single redirector, perhaps like the below, or something else entirely?

```js
  redirector({
    category: 'build',
    route: {
      base: '',
      pattern: ':alias(jenkins|jenkins-ci)/s/:protocol(http|https)/:host+/:job+',
    },
    transformPath: ({ protocol, host, job }) =>
      `/jenkins/build/${protocol}/${host}/${job}`,
    dateAdded: new Date('2019-04-20'),
```

Comment (paulmelnikow) [reply services/jenkins/jenkins-build-redirect.service.js]: Ah, sorry!

These two redirects are basically doing the same thing; one is for /jenkins/s` and the other for `/jenkins-ci/s`. This would be clearer if the `transformPath` functions, which are identical, weren't duplicated. Could you define the `transformPath` function above like:

```js
function transformPath({ protocol, host, job }) {
  ...
}
```

Comment (calebcartwright) [reply services/jenkins/jenkins-build-redirect.service.js]: oh also `host should be `.../:host/...` without the `+`

Comment (calebcartwright) [reply services/jenkins/jenkins-build-redirect.service.js]: Got it, thanks for the clarification. Do you think it would be more clear with two redirector services (leveraging a single transform function), or combining them into a single redirector?

File: services/test-validators.js (hunk 2 above)
Comment (calebcartwright) [reply services/test-validators.js]: Renamed accordingly

File: services/jenkins/jenkins-coverage-redirector.service.js (hunk 7 above)
Comment (calebcartwright) [reply services/jenkins/jenkins-coverage-redirector.service.js]: Nice! For some reason I had it in my mind that `createServiceTester` couldn't be used for redirects but that's for deprecated services

File: services/jenkins/jenkins-build-redirect.service.js (hunk 6 above)
Comment (paulmelnikow) [reply services/jenkins/jenkins-build-redirect.service.js]: I _think_ two would be more performant if we do #3328, so thinking ahead, I'd be inclined to stick with that!

File: services/jenkins/jenkins-coverage-redirector.service.js (hunk 7 above)
Comment (calebcartwright) [reply services/jenkins/jenkins-coverage-redirector.service.js]: Done!

File: services/test-results.js (hunk 1 above)
Comment (calebcartwright) [reply services/test-results.js]: Yeah I think having the full route in the doc was more helpful back before the window we have today that will auto populate the the query params for users. 

I can remove that section (and change this to a simple static prop) if we're good with no longer including that

File: services/jenkins/jenkins-build-redirect.service.js [hunk 11]
```diff
@@ -2,15 +2,28 @@
...
 const { redirector } = require('..')
 
+const transformPath = ({ protocol, host, job }) =>
+  `/jenkins/build/${protocol}/${host}/${job}`
+
+const commonProps = {
+  category: 'build',
+  transformPath,
+  dateAdded: new Date('2019-04-20'),
+}
```
Comment (paulmelnikow) [inline services/jenkins/jenkins-build-redirect.service.js]: 👍

**Comment by calebcartwright** [comment]
One last commit coming with a change to the doc
